### ISSUE TITLE
BaseControllerDataProvider refactor

### ISSUE THREAD

**Issue by FejZa** [XRTK/XRTK-Core#432]
# XRTK - Mixed Reality Toolkit Feature Request

## Is your feature request related to a problem? Please describe

`BaseControllerDataProvider` is being used by all the controller data providers as a base. It seems though, it does not provide the base functionality needed. Except for my newly created `HandControllerDataProviders` none of the existing data providres make use of the `ActiveControllers` list the base data provider provides. That is because it seems in many cases a simple list will not be enough, instead most of the data providers require a dictionary where they can store a controller via `Key-Value`. See e.g. `LuminControllerDataProvider` and `OculusControllerDataProvider`.

## How would you classify your suggestion

- Architecture / Services

## Describe the solution you'd like

I'd like to suggest upgrading the base data provider with the following changes, to provide a solid base for all other controllers and reduce code redundancy:

- In `BaseControllerDataProvider` refactor
```
private readonly List<IMixedRealityController> activeControllers = new List<IMixedRealityController>();

        /// <inheritdoc />
        public IReadOnlyList<IMixedRealityController> ActiveControllers => activeControllers;
```

to
```
private readonly Dictionary<string, IMixedRealityController> activeControllers = new Dictionary<string, IMixedRealityController>();

        /// <inheritdoc />
        public IReadOnlyDictionary<string, IMixedRealityController> ActiveControllers => activeControllers;
```

- In all child implementations revmoe their custom ActiveControllers dictionary and make use of the base provided.
- Provide protected accessibility methods if needed to conventiently read and write Controllers to the dictionary.

**Comment by FejZa** [comment]
Adding that I belive `IMixedRealityHandController GetOrAddController(Handedness handedness)` which is implemented in every derived class has potential for a solid base implementation since it seems most of the code is the same in all controller data providers except for determing the type of the controller requested.

**Comment by SimonDarksideJ** [comment]
This was actually the focus on my PR to unify Controller lists to the Input System :D 
https://github.com/XRTK/XRTK-Core/tree/dev/detectedcontrollersupport
But I paused due to the work on the new Platform system & Input Actions which also touches on some similar areas.
Once those are merged, we can revisit this

**Comment by StephenHodgson** [comment]
I think I've also addressed this in some way with my [Input Action Refactoring PR](https://github.com/XRTK/XRTK-Core/pull/379) as well.

**Comment by StephenHodgson** [comment]
I think most of the derived types use a more specific ActiveController dictionary with a key that best suits the platform (i.e. Lumin Controller Data Provider uses `Dictionary<byte, LuminController>` and Windows Mixed Reality Data Provider uses the `Dictionary<uint, IMixedRealityController>`

I think the Active Controllers is an override to get the `IMixedRealityControllers` that are active, and a dictionary is not required here.

**Comment by FejZa** [comment]
I am all in for a specific ActiveController list or whatever in the derived types but then IMO the base should not define an `ActiveControllers` list at all. It's error prone and confusing. Since the base also defines e.g. `RemoveController()` that will remove a controller out of the base list. But in derived types you most likely didn't even add to that base list.

I've stumbled upon that multiple times. If a base implementation does not provide any value, it shouldn't be there or be abstract or be an interface.

In summary what I am saying is, currently it's some mixture solution and a decision should be made how to solve it and then adjust the `BaseControllerDataProvider` to adhere to it.

**Comment by StephenHodgson** [comment]
I agree the Input System should be the only place to get the list of active controllers.

**Comment by FejZa** [comment]
I like the sound of that.

**Comment by StephenHodgson** [comment]
I think the controller data providers are only doing this for internal implementation details to make it easier to facilitate the management of controllers but that should just be an implementation detail

**Comment by StephenHodgson** [comment]
Yeah we already have `IMixedRealityInputSystem.DetectedControllers` seems redundant to have them in the base controller data provider's interface as well if we're not using it.

**Comment by StephenHodgson** [comment]
@SimonDarksideJ maybe you can weigh in more on this?

**Comment by SimonDarksideJ** [comment]
That is what my PR to resolve the "Multiple controllers being detected" PR was about. It effectively did away with the ActiveController registry inside each Data provider in favour of exposing and using the registered controllers in the InputSystem.
I parked by original branch due to the incoming changes to platforms and input actions as they would completely interfere with the approach.

Essentially, it updated the Inputsystem registry to recognise type and hand and some helpers to be able to resolve them.  The only issue I was unable to solve at the time was that Unity Joystick and Generic Controller were constantly fighting with each other.

I will return to this at some point

**Comment by StephenHodgson** [comment]
So then what is the status of this issue?

**Comment by SimonDarksideJ** [comment]
I need to revisit it once the Platform and Input Actions refactor are done, as it crossed paths with both

**Comment by StephenHodgson** [comment]
This issue specifically. Should we close it in favor of the other?

**Comment by FejZa** [comment]
I can't find the other issue to reference it. We should at least mention it here and then I am fine with closing.
